### ISSUE TITLE
Bug (?) - ReactiveForms > FormArray, trackBy index > removeAt & subscriptions 

### ISSUE THREAD

**Issue by borriej** [angular/angular#21682]
Hi Angular,

I think i found a [x] bug, but I'm not sure.

I have a large Reactive Form, with a FormArray.. and I trackBy index. Each FormGroup in the array has subscriptions like:

```
  // get the rides so we can iterate over them
  get myRides(): FormArray {
    return this.ridesForm.get('rides') as FormArray;
  };
```

```
  trackByFn(index, item) {
    return index; // or item.id
  }
```

```
 // each ride that is added has subscriptions
  subscribeChanges(ride: FormGroup){

    ride['dateSub$'] = ride.get('date').valueChanges.subscribe(val => {
     });
    
    ride['distanceSub$'] = ride.get('distance').valueChanges.subscribe(val => {
    });

    ride['startCountSub$'] = ride.get('start_count').valueChanges.subscribe(val => {
    });

    ride['endCountSub$'] = ride.get('end_count').valueChanges.subscribe(val => {
    });
  }
```

now.. you can also delete FormGroup's using:

```
  // deletes a single ride
  public deleteRide(ride: FormGroup, i: number): void{
        this.unsubscribeChanges(ride);
       this.myRides.removeAt(i)
}
````

but it seems that the tracking of objects goes wrong.. the subscriptions below the index of your removed FormGroup do not fire. As if their binding is still on a 'old' index. But since you've deleted a FormGroup - let's say at position 3 of 0 to 6 - the new FormGroup at position 3 has no subscriptions.

Is it not allowed to put your Subs$ on a FormGroup? Or is there a bug in the tracking / refreshing of FormGroup's in the array?

**Comment by jotatoledo** [comment]
Add a reproduction of the issue

**Comment by borriej** [comment]
working on it, i think it has to do with 'sort' on the array somewhere. second

**Comment by borriej** [comment]
@jotatoledo https://stackblitz.com/edit/angular-l4v8pr

**Comment by borriej** [comment]
Also note that i had two really strange bugs in my real code losing reference after saving.

```
- There is no FormControl instance attached to form control element with path: 'rides -> 4 -> date'
```
```
```

**Comment by pepemiso16** [comment]
Is there a way we delete specific FormArray withou changing the initial index of active FormArray? For example, formGoupname were 0,1,2,3 when we delete 2 it should be 0,1,3 instead of 0,1,2

**Comment by Evo267** [comment]
I'm also having this issue, is there a workaround?

**Comment by manojvignesh** [comment]
I had same kind of issue where i was storing the index and removing using them thinking index will never change. I was wrong. Then i started to find the index before removing to get the current index like `this.formArray.removeAt(this.formArray.controls.findIndex(control => control === myControl))`. This solved my problem.

**Comment by manojvignesh** [comment]
@palerdot myControl is a FormControl instance which is present in formArray. To remove it, i find the index of it first using the findIndex and pass the index to the `formArray.removeAt`. So since i can't rely on index to remove a entry from FormArray, i store the reference to the formControl which i add to formArray. Now when i want to remove it, i find the index and use the index to remove it.

**Comment by pepemiso16** [comment]
@manojvignesh can you show me how did you handle it for example there's a series of buttons to remove specific index?

**Comment by manojvignesh** [comment]
@pepemiso16 Well i've a form array with  5 controls in a component and i pass the **control** to the child components which has a delete button. When i've to delete a control from child component, i use the above mentioned code to delete it.  I don't maintain the index as it's changes, i maintain the control instance/reference.

**Comment by pepemiso16** [comment]
@manojvignesh something like this one https://stackblitz.com/edit/angular-l4v8pr

**Comment by manojvignesh** [comment]
@pepemiso16 Sry, just realized that my scenario doesn't have trackBy that's why it was working for me. Same way in your code also removing trackBy fixes the issue. But i do under stand that that's not what you want. You might have to wait for the fix or remove trackBy.

**Comment by adam-marshall** [comment]
The `this.formArray.controls.findIndex(control => control === myControl)` works for me, but it can be particularly painful for nested groups of controls. E.g. finding the arrayIndex where the array contains a group, which contains a group, which contains the control.

**Comment by samherrmann** [comment]
I am experiencing this issue as well using Angular v13. Here is my minimalistic reproduction which only involves a `trackBy` function and `FormArray.removeAt()`:
https://stackblitz.com/edit/angular-ivy-bqwaqd?file=src/app/app.component.html

**Comment by skbeplus** [comment]
We are noticing the same behavior as @samherrmann with angular v13
`FormArray.removeAt()` always removes the last element no matter what index is passed.

**Comment by paulkugener** [comment]
Same behavior with angular v14

**Comment by JeanMeche** [comment]
Well, to me this is not a bug. The index is not well suited for the `trackBy` function.

By removing one element you're effectively telling angular the element didn't change since the key passed to trackBy didn't change (due to the shifting). 

If you don't have a unique ID, don't use `trackBy`.

**Comment by GeorgeSafwatz** [comment]
I thought adding formControlName directive would give the input some basic attributes specificly value attribute, which didn't happen.
Therefore, every input the formArray didn't have value, this lead to removing the correct input elements, but not updating the UI correctly. Upon adding value attribute, removing elements from fromArray is working as it supposed to be without any troubles with the UI.
I know it's silly, but it might help someone here.

**Comment by ClaireBd** [comment]
Still got the issue with angular v17. And, as `track` is mandatory with the new control flow syntax, this is quite annoying...

**Comment by docentYT** [comment]
6 year later and the problem persists. Angular v18 same thing with new control flow syntax.

**Comment by mehmetgunacti** [comment]
Same issue with v18.

I create an FormArray<FormControl<..>>, then loop through that array in the template and track by $index, but also subscribe to each FormControl. Then, when one of the FormControls is removed (by user action), the remaining (subscribed) FormControls and the $index don't match anymore, which leads to weird UI behavior.

If we could add one custom property, a unique 'tag', to AbstractControl, which could then be used in a @for loop for tracking (instead of $index), this would solve the problem.

**Comment by abhijit-chikane** [comment]
hey do we have any workaround?

**Comment by mehmetgunacti** [comment]
> hey do we have any workaround?
> 

No, I ended up creating a wrapper class with an ordinary array in it.

There are additions to the Forms Api in Angular v18, have yet to look into what's new.
Angular is constantly improving, I'm sure there will be a way of solving it without writing custom code.

**Comment by docentYT** [comment]
> hey do we have any workaround?

instead of
```html
@for (foo of bar; track $index) {}
```
use 
```html
<div *ngFor="let foo of bar; let $index=index;">
```

**Comment by BojanKogoj** [comment]
> > hey do we have any workaround?
> 
> instead of
> 
> ```
> @for (foo of bar; track $index) {}
> ```
> 
> use
> 
> ```
> <div *ngFor="let foo of bar; let $index=index;">
> ```

That's similar to doing 
```ts
@for (foo of bar; track foo) {}
```

which solves this issue.

**Comment by jonesprabu** [comment]
Hi, 

I faced a similar issue with formArray items. After understanding how trackBy works, I found that it needs to be provided with a unique ID to allow Angular's ngFor to correctly identify and remap elements from the DOM after the structure changes, such as when an item is removed from the array or reordered through drag-and-drop functionality.

![image](https://github.com/user-attachments/assets/34f4f9e3-a849-4178-a0f4-7ae01550580d)

Forked and updated the stakblitz provide above in the comment and see it is working there.. 

https://stackblitz.com/edit/angular-mhaf99?file=app%2Fapp.component.ts,app%2Fmock.ts

**Comment by su4g** [comment]
When can it be used normally? I am currently using V18.
